### PR TITLE
Feature/zephyr support

### PR DESCRIPTION
**Pull request by Cimex97**
Hey, 
This PR implements an interface for the Zephyr-RTOS.

The following strategies are currently supported:
- ThroughSerial
- DualUDP

Also, the configuration is available over the zephyr kconfig environment.

If the implementation doesn't fit your guideline, we can discuss it.

### CODE REVIEW

**Comment by gioblu** [comment]
Ciao @Cimex97 the code looks clean, happy to see it was not a huge hassle to add the support for zephyr.
Thank you for your valuable contribution, I suspect zephyr users will find PJON useful. There are minor code-style changes I will make after accepting your pull. Is this source extensively tested already?

Thank you again 👍

**Comment by Cimex97** [comment]
I tested it on several boards. Those boards are connected to a PJONRouter board on an RS485-Bus via ThrougSerial.
The RouterBoard communicate with an app on a Laptop over PJON DualUDP, that poll each board with the Request-Responde procedure.

The general system looks like that:
pjon_issue

I only have some troubles with high data rate. But I think this is still a configuration problem for the ThroughSerial Timings.

**Comment by gioblu** [comment]
Ciao @Cimex97, have you checked the length of the serial buffers? Maybe at a high data rate they get filled up because you don't read quickly enough. If that is the case reducing `TS_READ_INTERVAL` could help.

**Comment by Cimex97** [comment]
I don't have the `TS_READ_INTERVAL` configuration option, because my branch is based on the tag `12.1`. So should I rebase it onto your master branch?

**Comment by gioblu** [comment]
Sorry, the `ThroughSerial` strategy in master was substituted by 12.1's `ThroughSerialAsync`, which demonstrated to be more efficient in all cases.

Ciao @Cimex97 sure, so you will be sure it works with the next version (13.0) that will be soon released. Would be really helpful for me and other users to have at least an example for zephyr (I don't have familiarity with zephyr).

**Comment by gioblu** [comment]
@Cimex97 please see:
- https://github.com/gioblu/PJON/commit/61b55bc92757833a2a0f1128edd384e0164f301e
- https://github.com/gioblu/PJON/commit/6ba09629ba52c98c87d3cd1fa232b9392b13661b

**Comment by Cimex97** [comment]
I started a simple example and recognize, that I have troubles with the zephyr build system. 
I will let you know if I finished the steps.

**Comment by Cimex97** [comment]
Hey @gioblu, I changed a bit the Zephyr interface to fit your "only header" for PJON. I also update the cmake files and add a PingPong example that use the PJON strategy DualUDP. This sample is tested on the [nucleo_f429zi](https://docs.zephyrproject.org/latest/boards/arm/nucleo_f429zi/doc/index.html) board.

File: examples/ZEPHYR/local/DualUDP/PingPong/Transmitter/west.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        - name: tinycrypt
+          path: modules/crypto/tinycrypt
+          revision: 3e9a49d2672ec01435ffbf0d788db6d95ef28de0
+
+        - name: hal_stm32
+          revision: d1bc80d021f4ebc31f6e8b36f14b738cc26c7b03
+          path: modules/hal/stm32
+
+        - name: pjon
+          url: git@github.com:Cimex97/PJON.git
```
Comment (Cimex97) [inline examples/ZEPHYR/local/DualUDP/PingPong/Transmitter/west.yml:31]:
```suggestion
          url: git@github.com:gioblu/PJON.git
```

File: examples/ZEPHYR/local/DualUDP/PingPong/Transmitter/west.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+          revision: feature/zephyr_support
```
Comment (Cimex97) [inline examples/ZEPHYR/local/DualUDP/PingPong/Transmitter/west.yml:32]:
```suggestion
          revision: 13.0
```

**Review by Cimex97 — COMMENTED** [review]
When you merge this request, you have to change this lines:
*revision* should the tag of the current stable PJON version.

**Comment by gioblu** [comment]
Sorry @Cimex97, your pull happened to be in between few major changes, I should have been more specific: 
- `ThourghSerial` configuration constants are changed (because we now use the ThroughSerialAsync impl) so the menuconfig configuration needs to be updated
- https://github.com/gioblu/PJON/pull/366/files#diff-83aeef700a5a192d68e562d9ccfe3725R10 here at line 13 and 28 the constants `PJON_INCLUDE_TS` and `PJON_INCLUDE_DUDP` are not used anymore 
- Now to select which strategy (or strategies) you want to use you can just do `#include "PJONDualUDP.h"` and then use `PJONDualUDP` instead of `PJON<DualUDP>` see: https://github.com/gioblu/PJON/blob/master/src/PJONDualUDP.h
- So you can then change then the example accordingly here: https://github.com/gioblu/PJON/pull/366/files#diff-4cdaa0179862252461aa7ad1dd3af053R29 inclusion and instantiation
- The last thing is, if you can return -1 in the zephyr interface `serial_get_char` and `serial_put_char` in case of failure do it it helps

I would have done those minor changes myself but I do not have a testing environment ready. 
Thank you again for your support.

**Comment by Cimex97** [comment]
I will do it tomorrow. Should I rebase the branch onto master or merge master into my branch, what do you prefer?

**Comment by gioblu** [comment]
The merge may be a little more annoying but will keep track of what effectively happened, thank you very much @Cimex97 

I suggest you to disable the acknowledgement feature if the system operates in master-slave mode, that will for sure help with performance specially at high speed.

**Comment by Cimex97** [comment]
Sorry @gioblu, I have to do a couple of works done, before I can test it. But in the next weeks I have not enough time for that.

**Comment by gioblu** [comment]
Ok @Cimex97 don't be preoccupied, I will merge your pull and fix it myself, I would just ask you to give it a try as soon as you can :)

**Comment by gioblu** [comment]
Ciao @Cimex97 I have merged and update config and examples to work with master, would you give it a try?

**Comment by Cimex97** [comment]
Hey @gioblu,
of course, I will do it in the next week.

**Comment by gioblu** [comment]
Thank you very much @Cimex97

File: examples/ZEPHYR/local/DualUDP/PingPong/Transmitter/west.yml (hunk 2 above)
Comment (Cimex97) [inline examples/ZEPHYR/local/DualUDP/PingPong/Transmitter/west.yml:32]: You must change this to revision: master

File: zephyr/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,38 @@
+if(CONFIG_PJON)
+zephyr_interface_library_named(PJON)
+
+zephyr_compile_definitions_ifdef(CONFIG_PJON_MAX_PACKETS PJON_MAX_PACKETS=CONFIG_PJON_MAX_PACKETS)
+zephyr_compile_definitions_ifdef(CONFIG_PJON_PACKET_MAX_LENGTH PJON_PACKET_MAX_LENGTH=CONFIG_PJON_PACKET_MAX_LENGTH)
+
```
Comment (Cimex97) [inline zephyr/CMakeLists.txt:6]: Also you has to add these two lines to here:

```
zephyr_compile_definitions_ifdef(CONFIG_PJON_INCLUDE_PORT PJON_INCLUDE_PORT)
zephyr_compile_definitions_ifdef(CONFIG_PJON_INCLUDE_MAC PJON_INCLUDE_MAC)
```

**Comment by Cimex97** [comment]
Hey @gioblu, I tested it and it works both, ThroughSerial and DualUDP.

**Comment by gioblu** [comment]
Ciao @Cimex97 thank you very much, here is the commit with the fixes you suggest: 
https://github.com/gioblu/PJON/commit/940c929d26076edf92e7a7a15404ce57ff2b8406

Thank you again for your support!

**Comment by Cimex97** [comment]
Thank you. If you have some trouble with it, feel free to ask me :).

**Pull request merged**